tests/integration/ha_tests/test_self_healing.py

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import pytest
+from pytest_operator.plugin import OpsTest
+from tenacity import Retrying, stop_after_delay, wait_fixed
+
+from tests.integration.ha_tests.helpers import (
+    METADATA,
+    app_name,
+    count_writes,
+    fetch_cluster_members,
+    get_primary,
+    is_replica,
+    postgresql_ready,
+    secondary_up_to_date,
+    send_signal_to_process,
+    start_continuous_writes,
+    stop_continuous_writes,
+)
+from tests.integration.helpers import get_unit_address
+
+PATRONI_PROCESS = "/usr/local/bin/patroni"
+POSTGRESQL_PROCESS = "postgres"
+DB_PROCESSES = [POSTGRESQL_PROCESS, PATRONI_PROCESS]
+
+
+@pytest.mark.abort_on_fail
+@pytest.mark.ha_self_healing_tests
+async def test_build_and_deploy(ops_test: OpsTest) -> None:
+    """Build and deploy three unit of PostgreSQL."""
+    # It is possible for users to provide their own cluster for HA testing. Hence, check if there
+    # is a pre-existing cluster.
+    if await app_name(ops_test):
+        return
+
+    charm = await ops_test.build_charm(".")
+    async with ops_test.fast_forward():
+        await ops_test.model.deploy(
+            charm,
+            resources={
+                "postgresql-image": METADATA["resources"]["postgresql-image"]["upstream-source"]
+            },
+            num_units=3,
+            trust=True,
+        )
+        await ops_test.model.wait_for_idle(status="active", timeout=1000)
+
+
+@pytest.mark.ha_self_healing_tests
+@pytest.mark.parametrize("process", DB_PROCESSES)
+async def test_restart_db_process(
+    ops_test: OpsTest, process: str, continuous_writes, master_start_timeout
+) -> None:
+    # Locate primary unit.
+    app = await app_name(ops_test)
+    primary_name = await get_primary(ops_test, app)
+
+    # Start an application that continuously writes data to the database.
+    await start_continuous_writes(ops_test, app)
+
+    # Freeze the database process.
+    await send_signal_to_process(ops_test, primary_name, process, "SIGTERM")
```

Comment:
Is it worth trying this with SIGKILL as well?

Reply:
`SIGKILL` will be used in the kill db process test. That test wasn't implemented yet due to some issues on pebble.